src/vsr/README.md

```diff
@@ -0,0 +1,189 @@
+
+## Protocol: Normal
+
+Normal protocol prepares and commits requests (from clients) and sends replies (to clients).
+
+1. The client sends a `command=request` message to the primary. (If the client's view is outdated, the receiver will forward the message on to the actual primary).
+2. The primary converts the `command=request` to a `command=prepare` (assigning it an `op` and `timestamp`).
+3. Each replica (in a chain beginning with the primary) performs the following steps concurrently:
+    - Write the prepare to the WAL.
+    - Forward the prepare to the next replica in the chain.
+4. Each replica sends a `command=prepare_ok` message to the primary once it has written the prepare to the WAL.
+5. When a primary collects a [commit quorum](#quorums) of `prepare_ok`s _and_ it has committed all preceding prepares, it commits the prepare.
+6. The primary replies to the client.
+7. The backups are informed that the prepare was committed by either:
+    - a subsequent prepare, or
+    - a periodic `command=commit` message.
+
+```mermaid
+sequenceDiagram
+    participant C0 as Client
+    participant R0 as Replica 0 (primary)
+    participant R1 as Replica 1 (backup)
+    participant R2 as Replica 2 (backup)
+
+    C0->>R0: Request A
+
+    R0->>+R0: Prepare A
+    R0->>+R1: Prepare A
+    R1->>+R2: Prepare A
+
+    R0->>-R0: Prepare-Ok A
+    R1->>-R0: Prepare-Ok A
+    R0->>C0: Reply A
+    R2->>-R0: Prepare-Ok A
+```
+
+See also:
+
+  - [VRR](https://pmg.csail.mit.edu/papers/vr-revisited.pdf) §4.1
+
+## Protocol: Start-View-Change
+
+Start-View-Change (SVC) protocol initiates [view-changes](#protocol-view-change) with minimal disruption.
+
+Unlike the Start-View-Change described in [VRR](https://pmg.csail.mit.edu/papers/vr-revisited.pdf) §4.2, this protocol runs in both `status=normal` and `status=view_change` (not just `status=view_change`).
+
+1. Depending on the replica's status:
+    - `status=normal` + primary: If the replica has not received a `prepare_ok` recently (and it expects to), send a `command=start_view_change` to all replicas (including self).
+    - `status=normal` + primary/backup: If the replica has not received a quorum of [heartbeats](#protocol-ping-replica-replica), send a `command=start_view_change` to all replicas (including self).
+    - `status=view_change`: If the replica has not completed a view-change recently, send a `command=start_view_change` to all replicas (including self).
+2. (Periodically retry sending the SVC).
+3. If the replica receives a heartbeat quorum or changes views (respectively), stop the `command=start_view_change` retries.
+4. If the replica collects a [view-change quorum](#quorums) of SVC messages, transition to `status=view_change` for the next view. (That is, increment the replica's view and start sending a DVC).
+
+To ensure that an idle (but healthy) cluster does not change views, a `status=normal` primary periodically prepares a heartbeat message.
```

Comment:
Hm, we don not actually _prepare_ heartbeats, right? We just send pings.

Reply:
Ah, thanks. Dead ~code~ documentation from a previous version.